feat(protocol-engine): Add tryLiquidProbe to complement liquidProbe

SyntaxColoring opened the pull request:
# Overview

This goes towards EXEC-598. See that ticket for background on the problem.

In v8.0.0, we think the only place this problem can happen is the `liquidProbe` command. `InstrumentContext.detect_liquid_presence()` wants to catch `liquidProbe` errors and turn them into `False` returns.

Instead of making that kind of `try`/`except` possible *in general,* we're solving it here specifically for `liquidProbe` commands by just having a second `liquidProbe` variant, `tryLiquidProbe`, that returns a `result` with `z_position=null` instead of returning an `error`. In other words, it's a variant of `liquidProbe` where `liquidNotFound` errors do not enter recovery mode. This is consistent with the `verifyTipPresence`/`getTipPresence` approach, and it's potentially useful for JSON protocols.

# Test plan

Just unit tests, so far. Follow-up work will wire this up to `opentrons.protocol_api` and then we can test it through that.

# Changelog

Add a `tryLiquidProbe` command, as described above.

To do:

* [x] Resolve conflicts with #15619. I think that should merge first and then I'll resolve the conflicts in this PR.
* [ ] PE state updates for the new command. We apparently don't have any state updates for the *old* command either, which seems like a bug?
* [ ] PAPI-level updates. I'll probably do these in a separate PR.

# Review requests

See inline notes.

# Risk assessment

To do.

# To do

SyntaxColoring committed:
Minor liquidProbe command fixups.

SyntaxColoring committed:
Add tryLiquidProbe models and implementation.

SyntaxColoring committed:
Update command unions.

SyntaxColoring committed:
Update command schema 8.

SyntaxColoring committed:
Minor liquidProbe unit test fixups.

SyntaxColoring committed:
Update liquidProbe/tryLiquidProbe unit tests.

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/liquid_probe.py:
```diff
@@ -34,6 +35,7 @@
 
 
 LiquidProbeCommandType = Literal["liquidProbe"]
+TryLiquidProbeCommandType = Literal["tryLiquidProbe"]
```
The names `liquidProbe`/`tryLiquidProbe` are inconsistent with `verifyTipPresence`/`getTipPresence`. I really don't know what to do about this at this point, since it seems like it'd be annoying to rename `liquidProbe`, but I'm open to ideas.

sfoster1 replied:
Would it be annoying? I don't think anybody uses it right now. I think we could do either
- `VerifyLiquidPresence`/`GetLiquidPresence`, if we want to add a distinct command for get liquid height
- If we don't want to change `LiquidProbe`, `VerifyLiquidPresence` seems like a good standalone command name honestly. Probably would want its own file though.

SyntaxColoring commented on shared-data/command/schemas/8.json:
I'm updating schema 8 in-place because I guess [we've been doing that](https://gist.github.com/SyntaxColoring/c1e9c621345377c12134d79c866d5017). But I think officially, we're supposed to be doing new stuff in a new schema 9.

One purpose of putting new stuff in new schemas is that the server can reject JSON protocols that are too new for it, efficiently and with a helpful error message, like we do for Python protocols. But that only works if we remember to put new stuff in new schemas, and I guess we haven't remembered, oops.

I'm happy to do whatever here. I'm uncertain whether we want to keep trying the separate-schema thing or just give up on it.

sfoster1 replied:
Yeah. I'm honestly not sure.

I think we hoped to put new commands or at least new patterns of command params in new schemas; so far we definitely haven't. I think it would be nice to start doing it again but I also think on some level we want to avoid having weird update races... I'm interested in @jerader 's opinion here.

SyntaxColoring committed:
Merge branch 'edge' into second_probe_command

Merging for analysis snapshot test updates.

sfoster1 reviewed (approved):
TWO commands??? in the same file???

sfoster1 commented:
Re: state updates - this is implemented right now as essentially stateless, if I understand what you're getting at. We're not tracking anything from this and don't really need to until phase 2. I think whether we should do it right now or not really depends on your outlook about state: is it for inputs for future commands, or is it for outside inspection? Nothing really looks at a possible liquid-verified state, so does it need to exist?

SyntaxColoring commented:
> Re: state updates

I'm thinking of the "current well" state used for path planning.

SyntaxColoring committed:
Merge branch 'edge' into second_probe_command

Resolve conflicts in:
* api/src/opentrons/protocol_engine/commands/liquid_probe.py
* api/tests/opentrons/protocol_engine/commands/test_liquid_probe.py

SyntaxColoring committed:
Add re-exports.

SyntaxColoring committed:
Add a separate params type.

This is necessary because PAPI, for convenience, maps param types to command requests and expects that mapping to be 1:1. Hoisted by my own petard. :(
